lib/puppet/resource/capability_finder.rb

```diff
@@ -0,0 +1,61 @@
+#
+# A helper module to look a capability up from PuppetDB
+#
+# @todo lutter 2015-03-10: determine whether this should be based on
+# Puppet::Pops::Evaluator::Collectors, or at least use
+# Puppet::Util::Puppetdb::Http
+
+require 'net/http'
+require 'cgi'
+
+module Puppet::Resource::CapabilityFinder
+
+  # Look the capability resource with the given +type+ and +title+ up from
+  # PuppetDB.
+  def self.find(environment, cap)
+    http = Net::HTTP.new("localhost", 8080)
+    query = ["and", ["=", "type", cap.type.capitalize],
+                    ["=", "title", cap.title.to_s],
+                    ["=", "tag", "producer:#{environment}"]].to_json
+
+    Puppet.notice "Capability lookup #{cap}]: #{query}"
+    response = http.get("/pdb/query/v4/resources?query=#{CGI.escape(query)}",
+                        { "Accept" => 'application/json'})
+
+    json = response.body
+
+    # @todo lutter 2014-11-04: use just JSON
+    data = PSON.parse(json)
+    data.is_a?(Array) or raise Puppet::DevError,
+    "Unexpected response from PuppetDB when looking up #{cap}: " +
+      "expected an Array but got #{data.inspect}"
+
+    # @todo lutter 2014-11-13: this was in the original capabilities
+    # prototype; can we really get entries back that do not have
+    # parameters set ?
+    data = data.select { |hash| hash["parameters"] }
```

Comment:
Changed to `map`. Added a comment pointing to [PuppetDB docs](http://docs.puppetlabs.com/puppetdb/3.0/api/query/v4/resources.html#response-format) and cleaned up the mess of how this was handled